## Project is about to predict human activity (Sitting,Laying etc). 
*   Data is colleted through sensor of 30 peopel doing differet activities.


In [369]:
import pandas as pd


In [370]:
data1=pd.read_csv('/content/drive/MyDrive/jmm_email_data/machine_learning_course_proj/human_activity_prediction/data1.csv')
data2=pd.read_csv('/content/drive/MyDrive/jmm_email_data/machine_learning_course_proj/human_activity_prediction/data1.csv')

In [371]:
data1.shape

(7352, 563)

In [372]:
data1.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject', 'Activity'],
      dtype='object', length=563)

In [373]:
data2.columns


Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject', 'Activity'],
      dtype='object', length=563)

#####Checking whether all columns names are same

In [374]:


(data1.columns==data2.columns).sum()

563

#####Combining the two datasets *vertically*

In [375]:
# Combining the two datasets *vertically*
combined=pd.concat([data1,data2],axis=0)

In [376]:
combined.shape

(14704, 563)

In [377]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14704 entries, 0 to 7351
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 63.3+ MB


#####reset index and drop the old index


In [378]:
# 
combined.reset_index(drop=True,inplace=True)

In [379]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14704 entries, 0 to 14703
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 63.2+ MB


**Artificial Intelligence**


*   input + rules = output -> Classical A.I (example-use of if-else condition is Classical A.I )
*   input + output = rules -> Machine Learning -> A to B mapping -> Supervised Learning ==>why we call it superwise because we are supervising the machine to learn A to B map, we are giving input as well as output.



Now machine has learnt the rule , when we give input machine has rules then it will give you output same as Classical A.I


*   input + rules = output



when training complete then we just give input and it will give resultt/output

In [380]:
combined.info

<bound method DataFrame.info of        tBodyAcc-mean()-X  tBodyAcc-mean()-Y  ...  subject          Activity
0               0.288585          -0.020294  ...        1          STANDING
1               0.278419          -0.016411  ...        1          STANDING
2               0.279653          -0.019467  ...        1          STANDING
3               0.279174          -0.026201  ...        1          STANDING
4               0.276629          -0.016570  ...        1          STANDING
...                  ...                ...  ...      ...               ...
14699           0.299665          -0.057193  ...       30  WALKING_UPSTAIRS
14700           0.273853          -0.007749  ...       30  WALKING_UPSTAIRS
14701           0.273387          -0.017011  ...       30  WALKING_UPSTAIRS
14702           0.289654          -0.018843  ...       30  WALKING_UPSTAIRS
14703           0.351503          -0.012423  ...       30  WALKING_UPSTAIRS

[14704 rows x 563 columns]>

In [381]:
##splitting dependent and independent variable

output=combined['Activity'].copy()
input=combined.drop(['Activity','subject'],axis=1).copy()

#####Standardizing the data

In [382]:
from sklearn.preprocessing import StandardScaler
slc=StandardScaler()
inp=slc.fit_transform(input)

#####Doing dimensionalit reduction ==>PCA


*   we have 563 variables (columns) in our data. Some of them might not contribute to give out the best result from the model.
*   Applying PCA to reduce the dimensions from 563 to 65.



In [383]:
from sklearn.decomposition import PCA

In [384]:
pca=PCA(n_components=65)

In [385]:
feature=pca.fit_transform(inp)
feature.shape

(14704, 65)

In [386]:
feature

array([[-16.13854371,   2.15202416,   3.14476896, ...,  -0.70590723,
          0.43351537,  -0.73337876],
       [-15.2961943 ,   1.38714392,  -0.68223056, ...,  -0.49901061,
          0.1053264 ,  -0.90345192],
       [-15.13701861,   2.47335077,  -1.75664537, ...,  -0.4998221 ,
          0.91448542,   0.60268512],
       ...,
       [ 14.33343587, -12.26071186,   4.02594548, ...,  -0.59865571,
          0.42327685,   0.16044217],
       [ 12.87601895, -14.07125618,   2.9160558 , ...,  -1.09566328,
         -0.22621925,   0.03526786],
       [ 13.01610365, -12.24426162,   1.33604195, ...,  -0.10650378,
          0.31247329,  -0.31933462]])

In [387]:
from sklearn.model_selection import train_test_split

In [388]:
x_train,x_test,y_train,y_test=train_test_split(feature,output,test_size=0.2,random_state=0)

In [389]:
print(x_train.shape,x_test.shape,y_train.shape,y_test.shape)

(11763, 65) (2941, 65) (11763,) (2941,)


In [390]:
output.unique()

array(['STANDING', 'SITTING', 'LAYING', 'WALKING', 'WALKING_DOWNSTAIRS',
       'WALKING_UPSTAIRS'], dtype=object)

#####USing KNN model to predict the activity

> Indented block



In [391]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import  classification_report,accuracy_score,confusion_matrix

In [392]:
knn=KNeighborsClassifier(5)   #5 neigbours
knn.fit(x_train,y_train)

KNeighborsClassifier()

In [393]:
predict=knn.predict(x_test)
predict[:5]

array(['WALKING', 'STANDING', 'LAYING', 'STANDING', 'WALKING'],
      dtype=object)

In [394]:
y_test[:5]

8771      WALKING
14536    STANDING
1524       LAYING
11546    STANDING
1091      WALKING
Name: Activity, dtype: object

In [395]:
(y_test==predict).sum()/2060

1.387378640776699

In [396]:
print(accuracy_score(y_test,predict))

0.9717783066984019


In [397]:
print(confusion_matrix(y_test, predict))


[[593   0   0   1   0   0]
 [  0 475  52   0   0   0]
 [  0  18 493   0   0   0]
 [  0   0   0 492   1   1]
 [  0   0   0   3 391   7]
 [  0   0   0   0   0 414]]


In [398]:
classification_report(y_test,predict)

'                    precision    recall  f1-score   support\n\n            LAYING       1.00      1.00      1.00       594\n           SITTING       0.96      0.90      0.93       527\n          STANDING       0.90      0.96      0.93       511\n           WALKING       0.99      1.00      0.99       494\nWALKING_DOWNSTAIRS       1.00      0.98      0.99       401\n  WALKING_UPSTAIRS       0.98      1.00      0.99       414\n\n          accuracy                           0.97      2941\n         macro avg       0.97      0.97      0.97      2941\n      weighted avg       0.97      0.97      0.97      2941\n'

#####USing LogisticRegression model to predict the activity




In [399]:
from sklearn.linear_model import LogisticRegression


In [410]:
log=LogisticRegression()
logist=log.fit(x_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [401]:
predicted=log.predict(x_test)

In [402]:
(y_test==predicted).sum()

2845

In [403]:
accuracy_score(y_test,predicted)

0.9673580414824889

In [404]:
confusion_matrix(y_test,predicted)

array([[594,   0,   0,   0,   0,   0],
       [  0, 474,  53,   0,   0,   0],
       [  0,  36, 475,   0,   0,   0],
       [  0,   0,   0, 491,   2,   1],
       [  0,   0,   0,   0, 400,   1],
       [  0,   0,   0,   2,   1, 411]])

In [405]:
test=y_test[:10]
pred=predict[:10]

In [406]:
actual_res=y_test.iloc[0]

In [407]:
actual_res

'WALKING'

In [408]:
predicted_result=knn.predict(x_test)[0]

In [409]:
predicted_result

'WALKING'